[shipy] - use docker-py to run commands instead of docker binary - please review

containscafeine opened the pull request.

cdrage commented:
@containscafeine I think what @dustymabe meant was keeping shipy as a submodule within Atomic App, right? https://git-scm.com/docs/git-submodule

containscafeine commented:
@cdrage nope, IIRC it was about copying the code to atomicapp ATM, instead of using a submodule, because they are messy! I have done the same using [git subtrees](http://blogs.atlassian.com/2013/05/alternatives-to-git-submodule-git-subtree/).

cdrage commented:
#dotests

cdrage commented on atomicapp/providers/lib/shipy/parser.py:
```diff
@@ -0,0 +1,414 @@
+    #     dest='restart_policy',
+    #     help='Restart policy to apply when a container exits '
+    #          '(no, on-failure[:max-retry], always, unless-stopped)'
+    # )
```
why isn't this added?

cdrage commented:
@containscafeine 

Just wondering as to why not all of the CLI params are added and some are commented out?

containscafeine commented:
@cdrage I have not added all because I have not written tests for them, yet. I have added all the frequently used ones, and will keep on adding other options later on.
Let me know about the params which I have skipped and which you think are necessary for the merge.

cdrage commented:
#dotests

cdrage commented on Makefile:
```diff
@@ -15,7 +15,7 @@ install:
 	pip install -qr requirements.txt
 	pip install -qr test-requirements.txt
-	$(PYTHON) -m pytest -vv
+	$(PYTHON) -m pytest tests/ -vv
```
may be better to add this in a separate PR :)

containscafeine replied:
Done

cdrage commented on atomicapp/providers/docker.py:
```diff
@@ -53,37 +55,49 @@ def init(self):
+            compatibility = {
+                '1.21': ('1.6.0', '1.7.0', '1.7.1', '1.7.2'),
+                '1.22': ('1.8.0', '1.8.1'),
+                '1.23': ('1.8.0', '1.8.1')
```
no support for 1.11.2, 1.10, etc?

containscafeine replied:
That is `{docker api version: docker-py version}`, and IIRC the latest docker API version is 1.23, so every docker version which uses 1.21, 1.22, 1.23 should be fine.

cdrage commented:
What happens when you pass a bad param? For example --link isn't supported anymore.

Error / warn out?

Also, lot's of debug output for Docker, is this normal (i'm asumming yes):
```
INFO:root:Name: agitated_ramanujan, ID: 976e1677
INFO:root:Name: hopeful_booth, ID: 86580ba8
INFO:root:Name: elated_mirzakhani, ID: 5dbefbc2
INFO:root:Name: elegant_allen, ID: d0833d82
INFO:root:Name: ecstatic_jang, ID: 2aff8bb1
INFO:root:Name: admiring_hawking, ID: 6e828e2f
INFO:root:Name: grave_euclid, ID: aad00e1f
INFO:root:Name: clever_albattani, ID: a524a37a
INFO:root:Name: goofy_brahmagupta, ID: 85627e45
INFO:root:Name: modest_volhard, ID: c9116eed
INFO:root:Name: furious_jang, ID: ce3ab880
INFO:root:Name: angry_ptolemy, ID: c102181e
INFO:root:Name: pensive_easley, ID: 95c0bfec
INFO:root:Name: mad_euclid, ID: fbd5e5d5
INFO:root:Name: elegant_bardeen, ID: b23d7824
INFO:root:Name: trusting_brattain, ID: b4a614cb
INFO:root:Name: cocky_dijkstra, ID: 74a46e09
INFO:root:Name: dropbox, ID: 277cb5fe
```

cdrage commented:
The debug info also lists containers which are currently offline ^^

containscafeine commented:
@cdrage 

> What happens when you pass a bad param? For example --link isn't supported anymore.

If docker-py can take it, it works. So the latest docker-py does take in config for `--link`, so it works. If it does not work, docker-py will error out, and so will shipy.

cdrage commented:
https://github.com/projectatomic/nulecule-library/tree/master/wordpress-centos7-atomicapp doesn't seem to fail despite Docker provider passing in `--link`?

cdrage commented:
Ahhhh that makes sense ^^

containscafeine commented:
@cdrage 
> The debug info also lists containers which are currently offline ^^

How did you get this? Did you run shipy standalone? Stopped containers should only be logged when `shipy ps -a` is run.

cdrage commented:
Running: `atomicapp run projectatomic/helloapache -v --provider=docker`

Producing:
```
DEBUG  :: - plugin.py :: Configuration option 'provider-config' not provided
DEBUG:atomicapp:Configuration option 'provider-config' not provided
DEBUG  :: - providers/docker.py :: Given config: {u'db_user': 'test', u'image': u'wordpress', 'namespace': 'default', u'db_pass': 'test', u'db_name': 'test', u'db_host': u'mariadb:3306', u'provider': u'kubernetes', u'port': 8080}
DEBUG:atomicapp:Given config: {u'db_user': 'test', u'image': u'wordpress', 'namespace': 'default', u'db_pass': 'test', u'db_name': 'test', u'db_host': u'mariadb:3306', u'provider': u'kubernetes', u'port': 8080}
DEBUG  :: - providers/docker.py :: Namespace: default
DEBUG:atomicapp:Namespace: default
INFO:root:
docker-py: 1.8.1
Client API: 1.22
Server API: 1.23
Compatibility: True
INFO   :: - providers/docker.py :: Deploying to provider: Docker
INFO:atomicapp:Deploying to provider: Docker
INFO:root:Name: mariadb-atomicapp-app, ID: 6e818aae
INFO:root:Name: default_centos-httpd_a1e313994835, ID: 95b00e66
INFO:root:Name: jolly_fermat, ID: b3013e2c
INFO:root:Name: rh_vpn, ID: 8cb6d982
INFO:root:Name: sick_brahmagupta, ID: 46a201db
INFO:root:Name: pensive_wilson, ID: 9c76d52a
```

containscafeine commented:
@cdrage 
> Also, lot's of debug output for Docker, is this normal (i'm asumming yes):

Could you please suggest on how to set up logging for shipy, how much should be logged out? Should the library calling shipy be able to set logging level for shipy, or is it fine just logging on its own?

cdrage commented:
It outputs all containers

cdrage commented:
@containscafeine Shipy seems to be using a custom logger? The logging output is messy / not the same as the logging library within Atomic App.

containscafeine commented:
@cdrage 
Ah yes, that happens due to [this](https://github.com/containscafeine/atomicapp/blob/1f1a03101fe3e780d6178ecdc840662f9c593ff6/atomicapp/providers/docker.py#L82), which is how shipy logs out. I should let atomicapp set shipy's logging level, right?

cdrage commented:
@containscafeine Correct!

Shipy uses:
```
logging.info('Running container {}.'.format(cid))
```

Atomic App uses:
```
logger = logging.getLogger(LOGGER_DEFAULT)
logger.debug("Given config: %s", self.config)
```

containscafeine commented on atomicapp/providers/lib/shipy/dpyexec.py:
```diff
@@ -0,0 +1,433 @@
+            logging.basicConfig(stream=sys.stdout, level=logging.DEBUG)
+            logging.debug('Setting logging level to logging.DEBUG')
+        else:
+            logging.basicConfig(stream=sys.stdout, level=logging.INFO)
```
@cdrage this is how shipy logs right now. If `-v` is passed, then DEBUG, else INFO.
How do you suggest this should behave instead?

containscafeine replied:
Should the program calling shipy be able pass it's own custom logger and shipy starts using that, is that fine?

cdrage replied:
that work work best, passing a custom logger to Shipy to use. Eliminates the usage of having "atomicapp" code in Shipy and still keeping it a separate library. Remember to keep it as an optional paramter to pass

containscafeine replied:
Done!

cdrage commented:
#dotests

cdrage commented:
Still not passing the correct logger :(
```
DEBUG  :: - nulecule/container.py :: Removing docker container: /usr/bin/docker rm -f 84c35e12f03a535295de9c3f36b0433f3cdaff63e86ef24edc11fb369776a737
DEBUG  :: - plugin.py :: Configuration option 'provider-config' not provided
DEBUG  :: - plugin.py :: Configuration option 'provider-config' not provided
DEBUG  :: - providers/docker.py :: Given config: {u'image': u'centos/httpd', u'hostport': 80, 'namespace': 'default', u'provider': u'kubernetes'}
DEBUG  :: - providers/docker.py :: Namespace: default
INFO:root:
docker-py: 1.8.1
Client API: 1.22
Server API: 1.23
Compatibility: True
INFO   :: - providers/docker.py :: Deploying to provider: Docker
INFO:atomicapp:Deploying to provider: Docker
INFO:root:Name: hopeful_rosalind, ID: df326c81
INFO:root:Name: rh_vpn, ID: 5659e35e
INFO:root:Name: modest_ardinghelli, ID: 78542cf3
INFO:root:Name: dropbox, ID: ba777285
INFO:root:Name: grave_hawking, ID: ee5a07e5
INFO:root:Name: amazing_heisenberg, ID: c0f8a8ce
INFO:root:Name: ansible_web_1, ID: b343b641
INFO:root:Name: clever_swartz, ID: 6910fd6e
INFO:root:Name: naughty_swanson, ID: 0fd81280
INFO:root:Name: adoring_ritchie, ID: ba69acce
INFO:root:Name: modest_hoover, ID: 7a507d79
INFO:root:Container image does not exist locally, pulling ...
INFO:root:No tag provided, pulling tag latest
^CINFO:root:Could not pull image centos/httpd:latest
```

cdrage commented:
@containscafeine also I think that *every* function should be implemented just-in-case :) have shipy as a complete library

containscafeine commented:
Hey @cdrage, updated logging and implemented all the commands, the only ones commented out now are for the rc release of docker-py, same for tests and flags too.

containscafeine commented:
Tests should pass once #775 is merged!

cdrage commented:
@containscafeine woo! thanks for all the hard work. I'll go through the code :)

cdrage commented:
Going to open/reclose in order to force-run unit tests again

cdrage commented:
#dotests

cdrage commented:
@containscafeine Can you rebase so the tests pass? I merged the other PR.

containscafeine commented:
@cdrage merged!

dustymabe commented on atomicapp/providers/docker.py:
```diff
@@ -54,37 +55,49 @@ def init(self):
-                       "on host (%s). Please update your host." % (client, server))
-                raise ProviderFailedException(msg)
+
+            version = self.shipy(cmd_check)
```
is a string returned. so `version` is just a string?

dustymabe commented on atomicapp/providers/docker.py:
```diff
@@ -54,37 +55,49 @@ def init(self):
+
+            version = self.shipy(cmd_check)
+
+            compatibility = {
```
would be great to not have to hardcode this. we should support whatever version docker-py supports right?

dustymabe commented on atomicapp/providers/docker.py:
```diff
@@ -54,37 +55,49 @@ def init(self):
+            containers = []
+
+            output = self.shipy(docker_cmd.split())
+            for container in output:
```
can we do `for line in output` instead. It's just a string of output correct? or is it more than that?

concaf commented:
Closing since irrelevant due to inactivity now.
